elements/includes/creative_commons.inc

```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+  // Get the path for the specified license.
+  $path = xml_form_elements_creative_commons_get_license_path($license);
+  // Return the specified license URI if the license or its path are unknown.
+  if (!in_array($license, $supported_licenses) || $path === FALSE) {
+    return $values['license_fieldset']['license_uri'];
+  }
+  // Get the license arguments for 'by'.
+  $arguments = '';
+  if ($license == 'by') {
+    if ($values['license_fieldset']['allow_commercial'] == 'n') {
       $arguments = "$arguments-nc";
     }
-    if ($derivatives == 'n') {
+    if ($values['license_fieldset']['allow_derivatives'] == 'n') {
       $arguments = "$arguments-nd";
     }
-    elseif ($derivatives == 'sa') {
+    elseif ($values['license_fieldset']['allow_derivatives'] == 'sa') {
       $arguments = "$arguments-sa";
     }
-    if ($jurisdiction == 'international') {
-      $license_version = '4.0';
-      // Not needed for international.
+  }
+  // Get the specified jurisdiction.
+  $jurisdiction = $values['license_fieldset']['license_jurisdiction'];
+  // Clean up jurisdiction if non-empty.
+  if (!empty($jurisdiction)) {
+    // Get the supported jurisdictions (assuming the first one should map
+    // to '').
+    $supported_jurisdictions = array_keys(xml_form_elements_creative_commons_jurisdiction_options());
+    // Empty jurisdiction if not supported or first (default) choice; otherwise
+    // append a slash.
+    if (!in_array($jurisdiction, $supported_jurisdictions) || $jurisdiction == $supported_jurisdictions[0]) {
       $jurisdiction = '';
     }
     else {
-      // If not international add the trailing slash.
       $jurisdiction .= '/';
     }
-    return "http://creativecommons.org/licenses/by$arguments/$license_version/$jurisdiction";
   }
+  // Get the specified version.
+  $version = $values['license_fieldset']['license_version'];
+  // If no version was specified, default to versions that should be available.
+  if (empty($version)) {
+    $version = ($license == 'by') ? ($jurisdiction == '' ? '4.0' : '2.5') : '1.0';
+  }
+  // Return the license URI.
+  return XML_FORM_ELEMENTS_CREATIVE_COMMONS_BASE_URL . "$path/$license$arguments/$version/$jurisdiction";
+}
+
+/**
+ * Process input and get values for `creative_commons` form element.
+ *
+ * @param array $element
+ *   The `creative_commons` form element.
+ * @param array|NULL $input
+ *   The user input (or NULL to retrieve defaults).
+ * @param array $form_state
+ *   The form state.
+ *
+ * @return array
+ *   The associative array representing the form element values.
+ *
+ * Note that validation for the `creative_commons` element is done whithin this
+ * function. This has the side effect that users cannot go to the previous step
+ * in a mulit-step form before resolving erroneous choices.
+ */
+function xml_form_elements_creative_commons_process_input(&$element, $input, &$form_state) {
```

Comment:
WHy does xml_form_elements_creative_commons_process_input require $input to be passed if $input comes from &$element and can be fetched from there?